packages/verified-fetch/README.md

```diff
@@ -658,6 +658,183 @@ Known Errors that can be thrown:
 4. `AbortError` - If the content request is aborted due to user aborting provided AbortSignal. Note that this is a `AbortError` from `@libp2p/interface` and not the standard `AbortError` from the Fetch API.
 
+## Pluggability and Extensibility
+
+Verified‑fetch can now be extended to alter how it handles requests by using plugins.
+Plugins are classes that extend the `BasePlugin` class and implement the `VerifiedFetchPlugin`
+interface. They are instantiated with `PluginOptions` when the `VerifiedFetch` class is created.
+
+Each plugin must implement two methods:
+
+- **`canHandle(context: PluginContext): boolean`**
+  Inspects the current `PluginContext` (which includes the CID, path, query, accept header, etc.)
+  and returns `true` if the plugin can operate on the current state of the request.
+
+- **`handle(context: PluginContext): Promise<Response | undefined>`**
+  Performs the plugin’s work. It may:
+  - **Return a final `Response`**: This stops the pipeline immediately.
+  - **Return `undefined`**: This indicates that the plugin has only partially processed the request
+    (for example, by performing path walking or decoding) and the pipeline should continue.
+  - **Throw a `PluginError`**: This logs a non-fatal error and continues the pipeline.
+  - **Throw a `PluginFatalError`**: This logs a fatal error and stops the pipeline immediately.
+
+Plugins are executed in a chain (a **plugin pipeline**):
+
+1. **Initialization:**
+   - The `VerifiedFetch` class is instantiated with a list of plugins.
+   - When a request is made via the `fetch` method, the resource and options are parsed to
+     create a mutable `PluginContext` object.
+
+2. **Pipeline Execution:**
+
+   - The pipeline repeatedly checks, up to a maximum number of passes (default = 3), which plugins
+     are currently able to handle the request by calling each plugin’s `canHandle()` method.
+   - Plugins that have not yet been called in the current run and return `true` for `canHandle()`
+     are invoked in sequence.
+     - If a plugin returns a final `Response` or throws a `PluginFatalError`, the pipeline immediately
+       stops and that response is returned.
+     - If a plugin returns `undefined`, it is assumed to have updated the context (for example, by
+       performing path walking), and the pipeline proceeds to a new pass where plugins are re‑evaluated.
+   - If no plugin modifies the context (i.e. no change to `context.modified`) and no final response is
+     produced after iterating through all plugins, the pipeline exits and a default “Not Supported”
+     response is returned.
+
+   **Diagram of the Plugin Pipeline:**
+
+flowchart TD
+A\[Resource & Options] --> B\[Parse into PluginContext]
+B --> C\[Plugin Pipeline]
+subgraph IP\[Iterative Passes max 3 passes]
+C1\[Check canHandle for each plugin]
+C2\[Call handle on ready plugins]
+C3\[Update PluginContext if partial work is done]
+C1 --> C2
+C2 --> C3
+end
+C --> IP
+IP --> D\[Final Response]
```

Comment:
The markdown generated here isn't rendering properly.. but it [renders correctly on the mermaid editor](https://mermaid.live/edit#pako:eNpVkk2PmzAQhv_KyIeeskmALslyqNQlrZpTUT8uhRwmeABrjY1ss5s04r_XQJq0-DLzzuP5whdWak4sYZXUb2WDxsGPXaHAfx_zb2R1b0qCd_C1c0Ire4CHhw_wnGdoLIFQTkMm-1qoVCtHJ3eYrz5PWJrPMchER1IoukZtf6wNdg3ss3zvyKATrwQZWksWWjxBBN3kXHmANMjThsoXKFF9QcUlQaUNEJYNdFONOxrmKUoJzYxpBYaQn6_YPymj_GfH0dH_A4CofHHjBEp40-YFhAWub62PrcyzhfeCsxDNAik-G-kk77PZ22eTu8s_C-VT-812fp10YAvWkmlRcP8PLiNbMNdQSwVLvMmpwl66ghVq8Cj2Tn8_q5IlzvS0YP00wU6g32f7V-xQseTCTizZroPl4-Z-4gU7s2Q9LNhvrT0fTPCvyZ7vGt3XDUsqlPaW_hMXTpubKDVyMmMFd-7Gl1ML63xvpVaVqEe9N9LLjXOdTVarMbyshWv647LU7coKPj6z5vUpXsVhvMUwongT4WMU8fIYPG2r8H1Q8c06CJENw_AHdDDVgg)

its late.. stepping away for now. i think i just need to wrap in ticks